Mathlib/RingTheory/AdicCompletion/Algebra.lean

```diff
@@ -0,0 +1,217 @@
+
+instance : Algebra R (AdicCompletion I R) :=
+  inferInstanceAs <| Algebra R (subalgebra I)
+
+@[simp]
+theorem val_one (n : ℕ) : (1 : AdicCompletion I R).val n = 1 :=
+  rfl
+
+@[simp]
+theorem val_mul (n : ℕ) (x y : AdicCompletion I R) : (x * y).val n = x.val n * y.val n :=
+  rfl
+
+/-- The canonical algebra map from the adic completion to `R ⧸ I ^ n`.
+
+This is `AdicCompletion.eval` postcomposed with the algebra isomorphism
+`R ⧸ (I ^ n • ⊤) ≃ₐ[R] R ⧸ I ^ n`. -/
+def evalₐ (n : ℕ) : AdicCompletion I R →ₐ[R] R ⧸ I ^ n :=
+  have h : (I ^ n • ⊤ : Ideal R) = I ^ n := by ext x; simp
+  AlgHom.comp
+    (Ideal.quotientEquivAlgOfEq R h)
+    (AlgHom.ofLinearMap (eval I R n) rfl (fun _ _ ↦ rfl))
+
+@[simp]
+theorem evalₐ_mk (n : ℕ) (x : AdicCauchySequence I R) :
+    evalₐ I n (mk I R x) = Ideal.Quotient.mk (I ^ n) (x.val n) := by
+  simp [evalₐ]
+
+/-- `AdicCauchySequence I R` is an `R`-subalgebra of `ℕ → R`. -/
+def AdicCauchySequence.subalgebra : Subalgebra R (ℕ → R) :=
+  Submodule.toSubalgebra (AdicCauchySequence.submodule I R)
+    (fun {m n} _ ↦ by simp; rfl)
+    (fun x y hx hy {m n} hmn ↦ by
+      simp only [Pi.mul_apply]
+      exact SModEq.mul (hx hmn) (hy hmn))
+
+/-- `AdicCauchySequence I R` is a subring of `ℕ → R`. -/
+def AdicCauchySequence.subring : Subring (ℕ → R) :=
+  Subalgebra.toSubring (AdicCauchySequence.subalgebra I)
+
+instance : CommRing (AdicCauchySequence I R) :=
+  inferInstanceAs <| CommRing (AdicCauchySequence.subring I)
+
+instance : Algebra R (AdicCauchySequence I R) :=
+  inferInstanceAs <| Algebra R (AdicCauchySequence.subalgebra I)
+
+@[simp]
+theorem one_apply (n : ℕ) : (1 : AdicCauchySequence I R) n = 1 :=
+  rfl
+
+@[simp]
+theorem mul_apply (n : ℕ) (f g : AdicCauchySequence I R) : (f * g) n = f n * g n :=
+  rfl
+
+/-- The canonical algebra map from adic cauchy sequences to the adic completion. -/
+@[simps!]
+def mkₐ : AdicCauchySequence I R →ₐ[R] AdicCompletion I R :=
+  AlgHom.ofLinearMap (mk I R) rfl (fun _ _ ↦ rfl)
+
+@[simp]
+theorem evalₐ_mkₐ (n : ℕ) (x : AdicCauchySequence I R) :
+    evalₐ I n (mkₐ I x) = Ideal.Quotient.mk (I ^ n) (x.val n) := by
+  simp [mkₐ]
+
+theorem Ideal.mk_eq_mk {m n : ℕ} (hmn : m ≤ n) (r : AdicCauchySequence I R) :
+    Ideal.Quotient.mk (I ^ m) (r.val n) = Ideal.Quotient.mk (I ^ m) (r.val m) := by
+  have h : I ^ m = I ^ m • ⊤ := by simp
+  rw [h, ← Ideal.Quotient.mk_eq_mk, ← Ideal.Quotient.mk_eq_mk]
+  exact (r.property hmn).symm
+
+theorem smul_mk {m n : ℕ} (hmn : m ≤ n) (r : AdicCauchySequence I R)
+    (x : AdicCauchySequence I M) :
+    r.val n • Submodule.Quotient.mk (p := (I ^ m • ⊤ : Submodule R M)) (x.val n) =
+      r.val m • Submodule.Quotient.mk (p := (I ^ m • ⊤ : Submodule R M)) (x.val m) := by
+  rw [← Submodule.Quotient.mk_smul, ← Module.Quotient.mk_smul_mk,
+    AdicCauchySequence.mk_eq_mk hmn, Ideal.mk_eq_mk I hmn, Module.Quotient.mk_smul_mk,
+    Submodule.Quotient.mk_smul]
+
+/-- Scalar multiplication of `R ⧸ (I • ⊤)` on `M ⧸ (I • ⊤)`. Used to have good
```

Comment:
Do you mean "In the past, it had", or "Designed to have"

Reply:
Ah right, it is not very clear as written. I want to say that we introduce this slightly weird `SMul` instance of `R ⧸ (I • ⊤)` on `M ⧸ (I • ⊤)` (instead of using the already existing `R ⧸ I`-module instance) to have good definitional behaviour.

Reply:
So maybe "This is used to" instead of "Used to"?